system/datatypes/mailinglist.php

```diff
@@ -0,0 +1,728 @@
+				$hyphaDictionary['ml_sender'] = 'afzender';
+				$hyphaDictionary['ml_email'] = 'e-mail';
+				$hyphaDictionary['ml_subscribe'] = 'inschrijven';
+				$hyphaDictionary['ml_successfully_subscribed'] = 'Succesvol ingeschreven';
+				$hyphaDictionary['ml_successfully_unsubscribed'] = 'Succesvol uitgeschreven';
+				$hyphaDictionary['ml_successfully_created'] = 'Succesvol aangemaakt';
+				$hyphaDictionary['ml_successfully_sent'] = 'Succesvol verstuurd';
+				$hyphaDictionary['ml_successfully_submitted_email'] = 'Succesvol verstuurd, bevestig de e-mail';
+				$hyphaDictionary['ml_status'] = 'status';
+				$hyphaDictionary['ml_draft'] = 'schets';
+				$hyphaDictionary['ml_sent'] = 'verstuurd';
+				$hyphaDictionary['ml_send'] = 'verstuur';
+				$hyphaDictionary['ml_addresses'] = 'adressen';
+				break;
+			default:
+			case 'en':
+				$hyphaDictionary['ml_editMailingList'] = 'edit mailing list';
+				$hyphaDictionary['ml_editMailing'] = 'edit mailing';
+				$hyphaDictionary['ml_createMailing'] = 'create a new mailing';
+				$hyphaDictionary['ml_successfully_updated'] = 'Successfully updated';
+				$hyphaDictionary['ml_subject'] = 'subject';
+				$hyphaDictionary['ml_sender'] = 'sender';
+				$hyphaDictionary['ml_email'] = 'email';
+				$hyphaDictionary['ml_subscribe'] = 'subscribe';
+				$hyphaDictionary['ml_successfully_updated'] = 'Successfully updated';
+				$hyphaDictionary['ml_successfully_subscribed'] = 'Successfully subscribed';
+				$hyphaDictionary['ml_successfully_unsubscribed'] = 'Successfully unsubscribed';
+				$hyphaDictionary['ml_successfully_created'] = 'Successfully created';
+				$hyphaDictionary['ml_successfully_sent'] = 'Successfully sent';
+				$hyphaDictionary['ml_successfully_submitted_email'] = 'Successfully sent, please confirm the email';
+				$hyphaDictionary['ml_status'] = 'status';
+				$hyphaDictionary['ml_draft'] = 'draft';
+				$hyphaDictionary['ml_sent'] = 'sent';
+				$hyphaDictionary['ml_send'] = 'send';
+				$hyphaDictionary['ml_addresses'] = 'addresses';
+				break;
+		}
+	}
+
+	public function indexAction() {
+		// check if form is posted and get form data
+		$formData = [];
+		$formPosted = $this->isPosted(self::FORM_CMD_LIST_SUBSCRIBE);
+		if ($formPosted) {
+			$formData = $_POST;
+		}
+
+		// create form
+		$form = $this->createSubscribeForm($formData);
+
+		// process form if it was posted
+		if ($formPosted) {
+			$form->validateRequiredField(self::FIELD_NAME_EMAIL);
+			$form->validateEmailField(self::FIELD_NAME_EMAIL);
+			if (empty($form->errors)) {
+				// check if email is already in the list, no need to have it twice.
+				$email = strtolower($form->dataFor(self::FIELD_NAME_EMAIL));
+				$addresses = $this->getDoc()->getOrCreate('addresses');
+				$address = $addresses->children()->findXPath('//address[@email="' . $email . '"]')->first();
+
+				if (!$address instanceof HyphaDomElement) {
+					$addAddress = true;
+					$sendEmail = true;
+					$code = uniqid();
```

Comment:
I believe `uniqid()` is based on the current timestamp and thus is fairly predictable. I the festival module, I used `bin2hex(openssl_random_pseudo_bytes(8))` to generate a more proper random identifier instead, which might also be useful here.

Reply:
In this case I don't really mind that the code is not unique or is predictable because I use it in combination with the email address, that being said, I have no objections to use your suggested method.
Note: `uniquid()` is currently used in `HyphaDomElement::generateId()` perhaps this should be updated as well, `uniquid` is not 100% unique which could cause problems.

Reply:
> In this case I don't really mind that the code is not unique
I guess that (with significant brute-forcing) subscribing someone else to a mailinglist is not really a big security problem, but it won't hurt to make it a bit more secure.

> or is predictable because I use it in combination with the email address
That does not really help, since an attacker would just subscribe someone else's e-mail address and then know the address.

>, that being said, I have no objections to use your suggested method.
Yes, please.

> Note: uniquid() is currently used in HyphaDomElement::generateId() perhaps this should be updated as well, uniquid is not 100% unique which could cause problems.

Correct, but that's inside a loop that checks for uniqueness, so that should result in a unique id eventually. Using a random id could reduce chances of conflicts, but I think those are already slim enough (and now the ids are sorted in order of creation, which can be useful).